books.rb

```diff
@@ -1,56 +1,63 @@
+require 'csv'
+require 'ostruct'
 require 'date'
 
-def to_date(date)
-    if date.length>4
-        Date.parse(date)
-    else
-        Date.parse("01-01"+date)
-    end
+def make_date(date)
+	if (date.size>4) 
+		Date.parse(date)
+	else (date.size<5)
+  	Date.parse("01-01"+date) 
+	end
 end
 
-def make_book_hash(book_array)
-  {
-    link: book_array[0],
-    title: book_array[1],
-    author: book_array[2],
-    pages: book_array[3].to_i,
-    date: to_date(book_array[4]),
-    rating: book_array[5].to_f,
-    genres: book_array[6]
-  }
+def make_openstruct(book_array)
+	{
+		link: book_array[0],
+		title: book_array[1], 
+		author: book_array[2], 
+		pages: book_array[3].to_i, 
+		date: make_date(book_array[4]), 
+		rating: book_array[5].to_f, 
+		genres: book_array[6]
+	}
 end
 
-books = File.read('books.txt')
-            .split("\n")
-            .map { |book_str| book_str.split('|') }
-            .map { |book_array| make_book_hash(book_array) }
-
+books = CSV.parse(File.read("books.txt"), row_sep: "\n", col_sep: "|")
+           .map { |book| OpenStruct.new(make_openstruct(book))}
 
 def cool_print(array_of_books)
```

Comment:
Wrong, this function should know how to print book, why there is iteration over array?